[Validator] Html5 Email Validation

PurpleBooth opened the pull request:
Currently we only support a very loose validation. There is now a
standard HTML5 element with matching regex. This will add the ability
to set a `mode` on the email validator. The mode will change the
validation that is applied to the field as a whole.

These modes are:

* loose: The pattern from previous Symfony versions (default)
* strict: Strictly matching the RFC
* html5: The regex used for the HTML5 Element

Deprecates the `strict=true` parameter in favour of `mode='strict'`

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #21531
| License       | MIT
| Doc PR        | symfony/symfony-docs#8487

<!--
- Bug fixes must be submitted against the lowest branch where they apply
  (lowest branches are regularly merged to upper ones so they get the fixes too).
- Features and deprecations must be submitted against the 3.4,
  legacy code removals go to the master branch.
- Please fill in this template according to the PR you're about to submit.
- Replace this comment by a description of what your PR is solving.
-->

PurpleBooth committed:
[Validator] Html5 Email Validation

Currently we only support a very loose validation. There is now a
standard HTML5 element with matching regex. This will add the ability
to set a `mode` on the email validator. The mode will change the
validation that is applied to the field as a whole.

These modes are:

* loose: The pattern from previous Symfony versions (default)
* strict: Strictly matching the RFC
* html5: The regex used for the HTML5 Element

Deprecates the `strict=true` parameter in favour of `mode='strict'`

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -24,13 +24,13 @@
+    private $mode;
 
-    public function __construct($strict = false)
+    public function __construct($mode = 'html5')
```
we need to make the change in a BC friendly way, so I think we cannot change the default behavior
true/false should still be handled as before as much as this makes sense

PurpleBooth replied:
Something along these lines? 

```php
    /**
     * @var string
     */
    private $mode;

    public function __construct($strict = false, $mode = 'html5')
    {
        $this->mode = $mode;

        if($strict === true) {
            $this->mode = Email::VALIDATION_MODE_STRICT;
        }
    }
```

stof replied:
No, we should have only 1 argument (as we don't want to keep a first unused argument in the future API)

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -52,33 +52,45 @@ public function validate($value, Constraint $constraint)
-                    ->setParameter('{{ value }}', $this->formatValue($value))
-                    ->setCode(Email::INVALID_FORMAT_ERROR)
-                    ->addViolation();
+                              ->setParameter('{{ value }}', $this->formatValue($value))
```
please use single indentation level as before, we don't align `->`

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -52,33 +52,45 @@ public function validate($value, Constraint $constraint)
+                              ->addViolation();
 
                 return;
-            } elseif (!interface_exists(EmailValidation::class) && !$strictValidator->isValid($value, false, true)) {
```
I would keep "strict" instead of "rfc" for maintenance purpose (will help when merging branches)

nicolas-grekas commented on src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php:
```diff
@@ -54,7 +54,6 @@ public function testExpectsStringCompatibleType()
     public function testValidEmails($email)
     {
         $this->validator->validate($email, new Email());
-
```
should be revert, to ease reviewing the patch

nicolas-grekas commented on src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php:
```diff
@@ -94,23 +93,23 @@ public function getInvalidEmails()
     }
 
-    public function testStrict()
+    public function testRfc()
```
same: let's keep strict everywhere to reduce the diff, thus help the review+future merges

nicolas-grekas reviewed (commented):
thanks for the PR!
some pure style comments for now to help next reviews
this PR should target 3.4 (rebase+github-retarget needed)

stof commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -52,33 +52,45 @@ public function validate($value, Constraint $constraint)
         }
 
-        if ($constraint->strict) {
+        var_dump($constraint->mode);
```
should be removed.

stof commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -52,33 +52,45 @@ public function validate($value, Constraint $constraint)
                 return;
             }
-        } elseif (!preg_match('/^.+\@\S+\.\S+$/', $value)) {
+        } elseif (Email::VALIDATION_MODE_LOOSE === $constraint->mode && preg_match('/^.+\@\S+\.\S+$/', $value)) {
```
this is broken. You removed the `!`, so considering any email *with* a `@` in it as invalid.

stof commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -52,33 +52,45 @@ public function validate($value, Constraint $constraint)
+                          ->addViolation();
+
+            return;
+        } elseif (Email::VALIDATION_MODE_HTML5 === $constraint->mode && preg_match(
```
missing negation too

stof reviewed (changes requested):
Note having any test for the new mode is wrong.

stof commented on src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php:
```diff
@@ -94,23 +93,23 @@ public function getInvalidEmails()
-    public function testStrict()
+    public function testRfc()
     {
-        $constraint = new Email(array('strict' => true));
```
Some tests must be kept with the existing name, to cover backward compatibility (if this is deprecated, it should be come a legacy test, but it should not disappear)

PurpleBooth replied:
I have a depreciation notice being raised for usage on the `strict` property. However this means the test that checks `strict` still works fails. What's the best thing to do?

stof commented on src/Symfony/Component/Validator/Constraints/Email.php:
```diff
@@ -31,8 +31,12 @@ class Email extends Constraint
     public $checkMX = false;
     public $checkHost = false;
-    public $strict;
+    public $mode = 'html5';
```
This is a BC break for 3 reasons:

- the `strict` property does not work anymore
- the default behavior has changed)
- the constraint now has a default mode, meaning it is not possible to configure a global default anymore at the config level (as the validator uses it only when the constraint does not have a config)

PurpleBooth commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -24,13 +24,17 @@
+        $this->defaultMode = $defaultMode;
+
+        if (true === $strict) {
+            $this->defaultMode = Email::VALIDATION_MODE_STRICT;
```
Not terribly happy with this. Suggestions?

nicolas-grekas replied:
something like that yes (sorry for the CS)
```
    public function __construct($defaultMode = Email::VALIDATION_MODE_STRICT)
    {
switch ($defaultMode) {
case VALIDATION_MODE_HTML5:
case VALIDATION_MODE_STRICT:
case VALIDATION_MODE_LOOSE: break;
default: $defaultMode = $defaultMode ? VALIDATION_MODE_STRICT : VALIDATION_MODE_LOOSE;
}
```

PurpleBooth commented:
I think I've addressed the feedback so far, excluding one where I have a question about depreciation notices.

Is there any additional feedback now that the diff is a little tidier?

xabbuh commented on UPGRADE-3.4.md:
```diff
@@ -411,6 +411,53 @@ Validator
    deprecated and will throw an exception in Symfony 4.0.
 
+ * The `strict` option of the `Email` constraint is deprecated and will
+   be removed in Symfony 4.0, use the `mode` option instead
```
The sentence should end with a dot.

xabbuh commented on UPGRADE-3.4.md:
````diff
@@ -411,6 +411,53 @@ Validator
+   private $property;
+   ```
+
+ * The constructing the `EmailValidator` with a boolean is deprecated and will
````
Calling the `EmailValidator` with a boolean constructor argument is deprecated [...]

xabbuh commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -23,14 +23,53 @@
+     * * Email::VALIDATION_MODE_STRICT
+     * * Email::VALIDATION_MODE_HTML5
+     *
+     * Using a boolean to set Email::VALIDATION_MODE_STRICT, or Email::VALIDATION_MODE_LOOSE is deprecated since version
```
Using a boolean argument to [...]

xabbuh commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -23,14 +23,53 @@
+     * Using a boolean to set Email::VALIDATION_MODE_STRICT, or Email::VALIDATION_MODE_LOOSE is deprecated since version
+     * 3.4 and will be removed in 4.0
+     *
+     * @param string|bool $defaultMode
```
We shouldn't document the possibility to pass boolean values anymore.

xabbuh commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -23,14 +23,53 @@
     {
-        $this->isStrict = $strict;
+        if (is_bool($defaultMode)) {
+            @trigger_error(
```
the whole call should be on one line

xabbuh commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -52,11 +91,25 @@ public function validate($value, Constraint $constraint)
-        if (null === $constraint->strict) {
-            $constraint->strict = $this->isStrict;
+        if (null !== $constraint->strict) {
+            @trigger_error(
```
the `trigger_error()` call should be on a single line

xabbuh replied:
And IMO we should trigger this deprecation when the constraint is constructed to be able to catch it earlier than at runtime.

PurpleBooth replied:
I've added this to the constructor of the Email constraint class now. I am not sure if I should leave this in there as well or not.

xabbuh commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -23,14 +23,53 @@
 class EmailValidator extends ConstraintValidator
 {
+    const PATTERN_HTML5 = '/^[a-zA-Z0-9.!#$%&\'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$/';
+    const PATTERN_LOOSE = '/^.+\@\S+\.\S+$/';
```
Should we tag the constants as internal?

ro0NL commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -78,13 +120,27 @@ public function validate($value, Constraint $constraint)
+                return;
+            }
+        } elseif (Email::VALIDATION_MODE_HTML5 === $constraint->mode) {
+            if (!preg_match(
```
single line? perhaps define pattern first to avoid the little code duplication.

ro0NL commented on src/Symfony/Component/Validator/Tests/Constraints/EmailTest.php:
```diff
@@ -0,0 +1,36 @@
+class EmailTest extends TestCase
+{
+    /**
+     * @group legacy
```
missing @expectedDeprecation?

ro0NL commented on src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php:
```diff
@@ -23,7 +23,19 @@ class EmailValidatorTest extends ConstraintValidatorTestCase
+    }
+
+    /**
+     * @group legacy
```
same

ro0NL reviewed (commented):
nice work :)

PurpleBooth commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -24,13 +24,56 @@
+            case Email::VALIDATION_MODE_LOOSE:
+                break;
+            default:
+                $defaultMode = Email::VALIDATION_MODE_LOOSE;
```
Should this be an exception maybe?

ro0NL commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -78,13 +131,15 @@ public function validate($value, Constraint $constraint)
-                ->setParameter('{{ value }}', $this->formatValue($value))
-                ->setCode(Email::INVALID_FORMAT_ERROR)
-                ->addViolation();
+        } elseif (array_key_exists($constraint->mode, self::$emailPatterns)) {
```
`elseif (isset(self::$emailPatterns[$constraint->mode]) && !preg_match(self::$emailPatterns[$constraint->mode], $value)) { }` ?

We still lack some mode normalization on the constraint (only default mode is normalized). Opens a small trap.

Also instead of normalizing.. i think i prefer an exception in case of unknown modes passed. Which seems the regular way of doing things; https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Cookie.php#L131

nicolas-grekas reviewed (approved):
would be great to have it in 3.4

ro0NL commented on src/Symfony/Component/Validator/Constraints/Email.php:
```diff
@@ -31,8 +35,32 @@ class Email extends Constraint
         self::HOST_CHECK_FAILED_ERROR => 'HOST_CHECK_FAILED_ERROR',
     );
 
+    public static $validationModes = array(
```
Perhaps mark @internal

ro0NL commented on src/Symfony/Component/Validator/Constraints/EmailValidator.php:
```diff
@@ -24,13 +24,51 @@
 
-    public function __construct($strict = false)
+    /**
+     * EmailValidator constructor.
```
The docblock is kinda obvious no? For me only @param would do.

PurpleBooth commented:
Failing tests come from current 3.4 branch (as of 98dae3edb1ba8fa35c06a8c9a459e0ffb9105826).

nicolas-grekas commented:
We talked about including this on 3.4 and decided it was too late. As such, moving to 4.1.

PurpleBooth commented:
I'll rebase against that tonight

PurpleBooth commented:
Okay that's done now. (broken tests from master, again)

nicolas-grekas commented:
Time to merge @symfony/deciders ?

xabbuh commented on src/Symfony/Component/Validator/Constraints/Email.php:
```diff
@@ -21,6 +21,10 @@
 {
+    const VALIDATION_MODE_HTML5 = 'html5';
+    const VALIDATION_MODE_STRICT = 'strict';
+    const VALIDATION_MODE_LOOSE = 'loose';
```
What about being explicit now and markt these constants as `public`?

xabbuh commented on src/Symfony/Component/Validator/Constraints/Email.php:
```diff
@@ -31,8 +35,37 @@ class Email extends Constraint
     public $checkHost = false;
+
+    /**
+     * @deprecated since version 3.4, to be removed in 4.0. Set mode to "strict" instead.
```
`@deprecated since version 4.1, to be removed in 5.0. Set mode to "strict" instead.`

xabbuh replied:
The other messages need to be updated accordingly.

xabbuh commented:
The deprecation also needs to be documented in `UPGRADE-4.1.md`, `UPGRADE-5.0.md` and the component's changelog file. But otherwise this looks good to me.

PurpleBooth commented:
I'll update those tonight, or at lunchtime if I get a chance 👍

PurpleBooth commented:
That should be the version numbers corrected in the deprication notices

fabpot reviewed (approved):
Looks great to me!

fabpot commented:
Thank you @PurpleBooth.
